Add TCPRoute and UDPRoute Support for L4 Load Balancing

Skcey opened the pull request:
### Proposed changes


Problem: 
NGF currently lacks support for TCPRoute and UDPRoute resources, which are essential for managing Layer 4 (TCP/UDP) traffic via the Gateway API.

Solution: 
This PR adds basic support for TCPRoute and UDPRoute to NGF, enabling Layer 4 load balancing. Key implementations include:
1. Added controllers to watch and process TCPRoute/UDPRoute resources, following the pattern used for HTTPRoute/TLSRoute.
2. Implemented route construction logic in the state graph to resolve listeners, backend services, and reference grants for L4 routes.
3. Extended the NGINX stream configuration generator to create upstream groups and server blocks for TCP/UDP traffic, mapping routes to their backend services.
4. Implemented support for multiple backend services in TCPRoute and UDPRoute, enabling traffic distribution to different backend services based on configured weights.

Testing: 
Packaged into an image and tested in my environment.

Please focus on (optional): If you any specific areas where you would like reviewers to focus their attention or provide
specific feedback, add them here.

Closes https://github.com/nginx/nginx-gateway-fabric/issues/4536

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [✔️ ] I have read the [CONTRIBUTING](https://github.com/nginx/nginx-gateway-fabric/blob/main/CONTRIBUTING.md) doc
- [ ✔️] I have added tests that prove my fix is effective or that my feature works
- [ ✔️] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [✔️ ] I have rebased my branch onto main
- [ ✔️] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### Release notes

If this PR introduces a change that affects users and needs to be mentioned in the [release notes](../blob/main/CHANGELOG.md),
please add a brief note that summarizes the change.

<!-- If this PR does not require a release note, you can just write NONE in the release-note block below. -->

```release-note
Add support for TCPRoute and UDPRoute
```

bjee19 commented:
Hey @Skcey, thanks so much for taking the effort in re-opening this and addressing the feedback. Most of the team is out on holiday, but will get back to review in the new year. However, could I ask for the old PR to instead be re-opened as having comment history helps with reviewing. Not too big of an issue if there are issues with it though.

Skcey commented:
> Hey @Skcey, thanks so much for taking the effort in re-opening this and addressing the feedback. Most of the team is out on holiday, but will get back to review in the new year. However, could I ask for the old PR to instead be re-opened as having comment history helps with reviewing. Not too big of an issue if there are issues with it though.

Happy New Year, hope you have a great holiday break!!! And I originally intended to reopen the old PR, but I found that maybe  I don't have  the  permission to do that

ciarams87 reviewed (approved):
I've tested this and everything is working as expected. We will follow up with documentation etc.

Amazing work @Skcey - we really appreciate all your effort and hard work on this! This is a really valuable contribution to the NGF project ❤️

sjberman commented on internal/controller/nginx/config/stream_servers.go:
```diff
@@ -2,8 +2,13 @@ package config
 
+	"github.com/go-logr/logr"
+
+	v1 "sigs.k8s.io/gateway-api/apis/v1"
```
```suggestion
	"github.com/go-logr/logr"
	v1 "sigs.k8s.io/gateway-api/apis/v1"
```

sjberman commented on internal/controller/state/dataplane/types.go:
```diff
@@ -98,18 +106,33 @@ type VirtualServer struct {
+	// UpstreamName refers to the name of the upstream that is used (single backend case).
 	UpstreamName string
+	// Upstreams holds multiple upstreams with weights (multi-backend case).
+	Upstreams []Layer4Upstream
```
To simplify this L4 virtual server, couldn't we just use the list and remove upstream name? For single backend cases, the list would just have one entry.

Skcey replied:
Since TLSRoute currently uses a single BackendRef and UpstreamName, I’m not very familiar with TLSRoute and didn’t want to modify its fields—so I kept these existing fields and added BackendRefs and Upstreams as supplements. 

Additionally, the reason TCPUpstreams mentioned wasn’t merged into StreamUpstreams is also due to the difference between multi-backend (for TCPRoute) and single-backend (for TLSRoute) designs, which makes it inconvenient to share many methods. 

If feasible, I can try removing the original BackendRef and UpstreamName and unifying the use of multi-backend fields instead.

sjberman commented on internal/controller/state/graph/route_common.go:
```diff
@@ -1336,3 +1440,176 @@ func getCookiePath(match v1.HTTPRouteMatch) string {
+	source           client.Object
+	refGrantResolver func(resource toResource) bool
+	namespace        string
+	routeType        string // "TCP" or "UDP"
```
Can we use the route type alias instead of regular string for this type?

Skcey replied:
ok

Skcey commented on internal/controller/state/graph/route_common.go:
```diff
@@ -115,13 +119,42 @@ type L4Route struct {
+// GetPrimaryBackendRef returns the primary backend reference.
+// For multi-backend routes, it returns the first BackendRef.
+// For single backend routes, it returns BackendRef.
+func (spec *L4RouteSpec) GetPrimaryBackendRef() BackendRef {
```
I sincerely apologize—this is indeed a duplicate method. The method that’s actually being used is GetBackendRefs(), and I will remove this redundant method immediately.

Skcey commented on internal/controller/nginx/modules/package-lock.json:
```diff
@@ -1639,7 +1637,6 @@
       "integrity": "sha512-E4t7DJ9pESL6E3I8nFjPa4xGUd3PmiWDLsDztS2qXSJWfHtbQnwAWylaBvSNY48I3vr8PTqIZlyK8TE3V3CA4Q==",
       "dev": true,
       "license": "MIT",
-      "peer": true,
```
I‘m sorry that I can’t quite recall why this file was modified. I will add it back and debug it further.

sjberman commented on internal/controller/nginx/config/maps.go:
```diff
@@ -68,8 +68,11 @@ func createStreamMaps(conf dataplane.Configuration) []shared.Map {
-		if u, ok := upstreams[server.UpstreamName]; ok && server.UpstreamName != "" && len(u.Endpoints) > 0 {
-			socket = getSocketNameTLS(server.Port, server.Hostname)
+		if len(server.Upstreams) > 0 {
+			upstreamName := server.Upstreams[0].Name
```
Can you leave a comment here that just says that TLSPassthroughServers currently only support a single backend?

sjberman reviewed (commented):
Awesome work!

Skcey commented on internal/controller/state/dataplane/configuration.go:
```diff
@@ -471,6 +554,9 @@ func buildServers(
 		}
+		if l.Source.Protocol == v1.TCPProtocolType || l.Source.Protocol == v1.UDPProtocolType {
+			continue
+		}
```
ok, finished
